numpyro/hmc_util.py

```diff
@@ -755,10 +766,16 @@ def single_chain_init(key, only_params=False):
         seeded_model = seed(model, key)
         model_trace = trace(seeded_model).get_trace(*model_args, **model_kwargs)
         constrained_values, inv_transforms = {}, {}
+        has_transformed_dist = False
         for k, v in model_trace.items():
             if v['type'] == 'sample' and not v['is_observed']:
-                constrained_values[k] = v['value']
-                inv_transforms[k] = biject_to(v['fn'].support)
+                if v['intermediates']:
+                    constrained_values[k] = v['intermediates'][0][0]
```

Comment:
nit: use `v['base_value']` instead of `v['intermediates'][0][0]`.

Reply:
The `base_value` key is only available when we substitute using `base_param_map`. How about we make `sample_with_intermediates` return a sample together with a dict with two keys:
`base_value` which holds the base value, and `transform_intermediates` which holds transform intermediates, where each transform intermediate is a list of `[output, logdet_intermediate]`? That way, we can set
```
constrained_values[k] = v['intermediates']['base_value']
```
and in the handler, we don't need to add the key `base_value` for our nodes.

Reply:
Sorry, I missed that we wouldn't have `base_value` populated here. I think your current design is reasonable in that case and we can stick with this. 

> and in the handler, we don't need to add the key base_value for our nodes.

I think we can do this even right now by using `intermediates[0][0]` as you are doing here (since `value` is populated inside `substitute` now and not split between `substitute` and `apply_stack`), but I am fine with keeping a redundant `base_value` in the trace dict if you think that's useful.